File: extractor-appender/appender-sql/src/main/java/ru/mydesignstudio/database/metadata/extractor/database/uk/UniqueKeyExtractor.java
package ru.mydesignstudio.database.metadata.extractor.database.uk;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.database.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.database.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extract.result.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.extract.result.UniqueKeyModel;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

@Slf4j
@Component
public class UniqueKeyExtractor implements TableMetadataAppender {
    @StringResource("classpath:sql/extract_unique_keys.sql")
    private String extractQuery;

    @Autowired
    private ExtractHelper helper;

    @Override
    public void append(TableMetadata metadata, String schemaName, String tableName) {
        log.info("Extracting unique keys for table {} in schema {}", tableName, schemaName);

        metadata.setUniqueKeys(helper.extract(extractQuery, new Object[]{tableName, schemaName}, UniqueKeyModel.class));

        log.debug("Extracted {} unique keys", metadata.getForeignKeys().size());
    }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/UniqueKeyExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.UniqueKeyModel

@Component
class UniqueKeyExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<UniqueKeyModel> {
        val query = queryProvider.provide("classpath:sql/extract_unique_keys.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
            UniqueKeyModel(
                    owner = getString(rs, "owner"),
                    constraintName = getString(rs, "constraint_name"),
                    tableName = getString(rs, "table_name"),
                    columnName = getString(rs, "column_name"),
                    uniqueKey = getString(rs, "unique_key")
            )
        }
    }
}